### ISSUE TITLE
document co-run zookeeper and bookie on same machine

### ISSUE THREAD

**Issue by sijie** [apache/pulsar#3213]
#### Expected behavior

zookeeper and bookie should be able to co-run on same machine

#### Actual behavior

bookie will fail to start because zookeeper and bookie are using same prometheus port

#### Steps to reproduce

- start zookeeper
- start bookie

#### System configuration
**Pulsar version**: 2.2

**Comment by foreversunyao** [comment]
@sijie  this case is only for container env? I have an env, but zk and bookie are running on the same  server?

**Comment by sijie** [comment]
@foreversunyao : 

it can happen at running zookeeper and bookie on one same node (it can be one docker, one vm or physical machine). 

> I have an env, but zk and bookie are running on the same server?

did you successfully run zk and bookie one same server?

**Comment by foreversunyao** [comment]
@sijie I have to say yes, but actually, I used the official zookeeper version, and I just tried the pulsar zk, it is not working due to the "Address already in use".

**Comment by foreversunyao** [comment]
@sijie is it possible let me give a short if I can fix it? : )

**Comment by sijie** [comment]
@foreversunyao - yeah. it is awesome that you are willing to contribute a fix.

**Comment by foreversunyao** [comment]
welcome, 

@sijie  technically it's not a big impact fix, but the **default port** for zk is in code(not configuration), if it is changed, the upgrade could be hard and have risks in the future?
or we could change the prometheus export port ?

they both don't have many files to chnage.

**Comment by sijie** [comment]
I was thinking - most of the people probably don't co-run zk and bk on the same node. so we don't necessarily need to change the default zk stats port. so a simple approach might be adding a note to the deployment guide (http://pulsar.apache.org/docs/en/deploy-bare-metal/#deploying-a-zookeeper-cluster) - something like: if you are planning to co-run zookeeper with bookie on the same node, you should pick up a different stats port on running zookeeper, for example:

```
PULSAR_EXTRA_OPTS="-Dstats_server_port=8001" bin/pulsar-daemon start zookeeper
```

**Comment by foreversunyao** [comment]
@sijie , sure . what about I made this parameter to configurable while I update the doc also ?


stats_server_port is like a mandatory parameter for zk in pulsar, so it could be better in the configuration(zookeeper.conf), instead of "hardcode" ?

**Comment by sijie** [comment]
@foreversunyao yeah, so good to me. +1

**Comment by foreversunyao** [comment]
@sijie 
I just made a pr , but only updating the doc without change the code.

The reason is after I checked the configs for current zookeeper(zookeeper.conf and global_zookeeper.conf), all the parameters are from the official zookeeper template, without any special parameter for pulsar. So it's hard to add a new **pulsar exclusive** parameter to the config.

And also looks like pulsar-zookeeper just does two things( http service for stats and official zookeeper), a little confused why don't have a dedicate component for monitor all other components in puslar, like burrow for kafka ?

**Comment by sijie** [comment]
> So it's hard to add a new pulsar exclusive parameter to the config.

Yeah. That was my original thought as well. so I think only updating the doc should be good enough for now.

> a little confused why don't have a dedicate component for monitor all other components in puslar, like burrow for kafka ?

If I understand burrow correctly, burrow is an http service that does consumer lag checks. What the pulsar-zookeeper does is providing stats for monitoring zookeeper service itself. The same component that serves same purpose as burrow in Pulsar is actually in broker itself. Broker exposes all topics, producers and consumers related metrics through its http endpoint.

**Comment by foreversunyao** [comment]
@sijie ,  you are right about burrow.

And what I am concerned about is if it could be a better way to isolate the http server for zk stats and zk service itself to different parts in code ,  If I understand the code (ZooKeeperStarter.java) correctly , it looks like this stats service is running or not should not affect zk service itself ?

**Comment by sijie** [comment]
@foreversunyao -

sorry, I missed your comment.

> it looks like this stats service is running or not should not affect zk service itself ?

Correct. The pulsar zookeeper is just a starter wrapper for adding additional stats service for collecting stats from zookeeper and expose to prometheus.

**Comment by sijie** [comment]
This task has been fixed by #3242
